pkg/component/controller/etcd_member_reconciler.go

```diff
@@ -0,0 +1,316 @@
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	// find the member ID for this node
+	etcdClient, err := etcd.NewClient(e.k0sVars.CertRootDir, e.k0sVars.EtcdCertDir, e.etcdConfig)
+	if err != nil {
+		return err
+	}
+
+	peerURL := fmt.Sprintf("https://%s:2380", e.etcdConfig.PeerAddress)
+
+	memberID, err := etcdClient.GetPeerIDByAddress(ctx, peerURL)
+	if err != nil {
+		return err
+	}
+
+	// Convert the memberID to hex string
+	memberIDStr := fmt.Sprintf("%x", memberID)
+
+	name, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+	var em *etcdv1beta1.EtcdMember
+
+	log.WithField("name", name).WithField("memberID", memberID).Info("creating EtcdMember object")
+
+	// Check if the object already exists
+	em, err = e.etcdMemberClient.Get(ctx, name, v1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Debug("EtcdMember object not found, creating it")
+			em = &etcdv1beta1.EtcdMember{
+				ObjectMeta: v1.ObjectMeta{
+					Name: name,
+				},
+				PeerAddress: e.etcdConfig.PeerAddress,
+				MemberID:    memberIDStr,
+				Leave:       false,
+			}
+			_, err = e.etcdMemberClient.Create(ctx, em, v1.CreateOptions{})
+			if err != nil {
+				return err
+			}
+			return nil
+		} else {
+			return err
+		}
+	}
+
+	em.PeerAddress = e.etcdConfig.PeerAddress
+	em.MemberID = memberIDStr
+	em.Leave = false
+
+	log.Debug("EtcdMember object already exists, updating it")
+	// Update the object if it already exists
+	em, err = e.etcdMemberClient.Update(ctx, em, v1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	em.Status = etcdv1beta1.Status{
+		JoinStatus:      etcdv1beta1.JoinStatusJoined,
+		ReconcileStatus: "",
+		Message:         "",
+	}
+	_, err = e.etcdMemberClient.UpdateStatus(ctx, em, v1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+const EtcdFinalizer = "etcd.k0sproject.io/etcd-reconciler"
+
+func (e *EtcdMemberReconciler) reconcileMember(ctx context.Context, member *etcdv1beta1.EtcdMember) {
+	log := logrus.WithFields(logrus.Fields{
+		"component":   "etcdMemberReconciler",
+		"phase":       "reconcile",
+		"name":        member.Name,
+		"memberID":    member.MemberID,
+		"peerAddress": member.PeerAddress,
+	})
+
+	if !e.leaderElector.IsLeader() {
+		log.Debug("not the leader, skipping reconcile")
+		return
+	}
+
+	var err error
+
+	log.Debugf("reconciling EtcdMember: %+v", member)
+
+	if !member.Leave {
+		log.Debug("member not marked for leave, no action needed")
+		return
+	}
+
+	if member.Status.JoinStatus == etcdv1beta1.JoinStatusLeft {
+		log.Debug("member already left, no action needed")
+		return
+	}
+
+	// TODO Do we need to verify the conditions for removing peer? Or rely on Etcd client to error out if a peer cannot be removed?
+	// TODO I need to split up the logic a bit so the error handling is easier to understand/handle
+	defer func() {
```

Comment:
I wonder if it would be clearer to have two functions. One calling the other, and then creating events depending on the outcome. Might be easier to follow mentally than by using a defer func.

Reply:
You mean something like this:
```
if err := doSomething(); err != nil {
    updateState(...)
    return
}
if err := doSomethingElse(); err != nil {
    updateState(...)
    return
}
```

I was trying to avoid that as there's so many error cases. It seemed "easier" to store the error state and handle it in defer.